core/sequencer/server.go

```diff
@@ -108,39 +109,85 @@ func NewServer(
 
 // RunBatch reads mutations out of the queue and calls CreateEpoch.
 func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.Empty, error) {
-	batch, err := s.queue.ReadQueue(ctx, in.DomainId, in.MaxBatch)
+	// Get the previous and current high water marks.
+	domain, err := s.ktServer.GetDomain(ctx, &ktpb.GetDomainRequest{DomainId: in.DomainId})
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "ReadQueue(): %v", err)
+		return nil, err
+	}
+	mapClient, err := tclient.NewMapClientFromTree(s.tmap, domain.Map)
+	if err != nil {
+		return nil, err
+	}
+	latestMapRoot, err := mapClient.GetAndVerifyLatestMapRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var lastMeta spb.MapMetadata
+	if err := proto.Unmarshal(latestMapRoot.Metadata, &lastMeta); err != nil {
+		return nil, err
+	}
+	high, err := s.queue.HighWatermark(ctx, in.DomainId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "HighWatermark(): %v", err)
+	}
+
+	// TODO(gbelvin): If time since last map revision > max timeout, run batch.
+	// TODO(#1047): If time since oldest queue item > max latency has elapsed, run batch.
+	// TODO(gbelvin): If count items > max_batch, run batch.
+
+	// Count items to be processed.  Unfortunately, this means we will be
+	// reading the items to be processed twice.  Once, here in RunBatch
```

Comment:
I'm not 100% what to do here.  Keeping in mind that we're working towards a flow that consists of the following steps:

0. RunBatch - Start this flow
1. CommitBatch
   a. Collect info about current high water marks. 
   b. Collect info about current map revision.
   a. Optionally write (multiple?) batch boundaries in the batch definitions table
2. ApplyBatch
   a. Read batch boundary definitions from table
   b. Read map values
   c. Apply mutations
   d. Write to map
3. PublishBatch
   a. Copy all SignedMapHeads to Trillian Log of SignedMapHeads


Given this flow (which we need for reproducibility reasons) getting rid of the double read isn't straight forward...
1.  If we get rid of count based epoch creation controls `min_batch` and `max_batch` we wouldn't need to read twice.
     a. `min_batch` can easily be replaced with `max_latency` #1047
     b. `max_batch` is harder -- maybe it can be replaced with a maximum time period between low and high? 
2. We can't have `CreateEpochFromMessages` because it wouldn't be reproducible.

Reply:
Yeah, good point, `CreateEpochFromMessages` would push the requirement for reproducability outwards -- safer to re-read.  (I guess we can re-visit if this turns out to be a bottleneck, but I suspect it won't be.)